src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,212 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. B. G. Chalk, STFC Daresbury Lab
+
+'''This module provides the sclarization transformation class.'''
+
+import itertools
+
+from psyclone.core import VariablesAccessInfo
+from psyclone.psyGen import Kern
+from psyclone.psyir.nodes import Assignment, Call, CodeBlock, IfBlock, \
+        Reference, Routine
+from psyclone.psyir.symbols import DataSymbol
+from psyclone.psyir.transformations.loop_trans import LoopTrans
+
+
+class ScalarizationTrans(LoopTrans):
+
+    def _find_potential_scalarizable_array_symbols(self, node, var_accesses):
+
+        potential_arrays = []
+        signatures = var_accesses.all_signatures
+        for signature in signatures:
+            # Skip over non-arrays
+            if not var_accesses[signature].is_array():
+                continue
+            # Skip over non-local symbols
+            base_symbol = var_accesses[signature].all_accesses[0].node.symbol
+            if not base_symbol.is_automatic:
+                continue
+            array_indices = None
+            scalarizable = True
+            for access in var_accesses[signature].all_accesses:
+                if array_indices is None:
+                    array_indices = access.component_indices
+                # For some reason using == on the component_lists doesn't work
+                elif array_indices[:] != access.component_indices[:]:
+                    scalarizable = False
+                    break
+            # For each index, we need to check they're not written to in
+            # the loop.
+            flattened_indices = list(itertools.chain.from_iterable(
+                    array_indices))
+            for index in flattened_indices:
+                sig, _ = index.get_signature_and_indices()
+                if var_accesses[sig].is_written():
+                    scalarizable = False
+                    break
+            if scalarizable:
+                potential_arrays.append(signature)
+
+        return potential_arrays
+
+    def _check_first_access_is_write(self, node, var_accesses, potentials):
+        potential_arrays = []
+
+        for signature in potentials:
+            if var_accesses[signature].is_written_first():
+                potential_arrays.append(signature)
+
+        return potential_arrays
+
+    def _check_valid_following_access(self, node, var_accesses, potentials):
+        potential_arrays = []
+
+        for signature in potentials:
+            # Find the last access of each signature
+            last_access = var_accesses[signature].all_accesses[-1].node
+            # Find the next access to this symbol
+            next_access = last_access.next_access()
+            # If we don't use this again then its valid
+            if next_access is None:
+                potential_arrays.append(signature)
+                continue
```

Comment:
I was expecting the check for next accesses to only be this. Why do we need the rest, is it because the next_access implementation is currently missing these? E.g. if we use next_accesses in other places, e.g. when deciding where to place a halo_exchage to be safe, wouldn't we need the same logic as below? In this case shouldn't this all be inside the next_access implementation?

Also I see that next access only returns one reference, was this the right choice?, given that potentially is more than one:
```
a = 3
if condition:
    a = 1
else:
    a = 1
```
Also the `a` in the inner loop has 2 potential next_acesses
```
do i=1, 10
     a = 1
     do j 1=1, 10
          a = 2
     endo
enddo
a = 3
```

Reply:
I deliberately kept the next_access function simple and tried to avoid the complicated cases (and rely on the dependency analysis tooling) like this, and just assume these are ok.

I can try to make a new issue and PR to improve that functionality if you think its necessary/useful. I feel like there can always be an incredibly complex structure of nested loops that makes this difficult.

Reply:
But wouldn't this complex logic be for everywhere that we want to prove if a symbol is going to be accessed next or not, not just for this transformation?

Reply:
> I deliberately kept the next_access function simple and tried to avoid the complicated cases (and rely on the dependency analysis tooling)

I am not sure if it's next access or the VariablesAccessInfo tooling itself that needs to provide this.

@hiker and @JulienRemy I am trying to understand this. Is this related to the logic you needed on building the DAG? That the VariablesAccessInfo returns ALL following dependencies under a node (but not looking upwards - like we need in my loop snippet above) and the next_accesses just one is code order?

But what we need (and maybe next_access should be) is a method that return any "directly reachable" accesses to the same symbol, which could be more than one, and will be equivalent to the arrows leaving from a node in the DAG?

Reply:
I think this is a separate issue (though I'm not sure about the loop case, we might need to gracefully fail somehow for some case, I'll try to remember to put a test in for that).

We otherwise ignore the if case here (which could be improved if next_acccess gave all options, either through the access info or otherwise) and return False instead of trying to work it out.

Reply:
@sergisiso I think I could probably do most of this for next access, butt I'm not sure on if there is an easy way to tell whether a node is inside an else if or an else or an if, for example for a case like (this is how we have the if/elseif behaviour in Psyclone)
```
a = 1
if cond1 then
   a = 2
else
  if cond2 then
    a = 3
    a = 6
  else
    if cond3 then
       a=4
end if 
end if
end if
a = 5
```

We currently would find the `a=2` case, and the proposal is that we need to also find the `a=3`, `a=4`, and `a=5` cases. The trick is knowing that the `a=3` and `a=4` is inside an else if case, so to continue trying to find more "next accesses" that aren't inside the same branch of the else/if structures. We also need different behaviour for `else if` vs `else` behaviour (and all cases are complicated by the "one branch doesn't contain a next_access")